### PR TITLE
meson: add meson support

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
Looks pretty good, thanks!

Can you split:
 - Code refactoring (I see some defines, includes, chunks, ... moving around)
 - Code changes needed because of stricter build flags
 - Removal of autoconf/automake
 - Addition of meson
 - Porting of CI scripts
 - Updating README and any other build instruction

Also, from our discussion yesterday I thought we wanted to keep it building with 16.04 because of our large remaining userbase there. So we'd need to re-introduce that Github Action chunk and make it behave.

That should make it easier to review as well as backport.
For stable-4.0, I'm tempted to have us offer both. That way we don't regress current build scripts but we also have meson for those who prefer it.

**Comment by brauner** [comment]
> Looks pretty good, thanks!
> 
> Can you split:
> 
> * Code refactoring (I see some defines, includes, chunks, ... moving around)
> * Code changes needed because of stricter build flags
> * Removal of autoconf/automake
> * Addition of meson
> * Porting of CI scripts
> * Updating README and any other build instruction
> 
> Also, from our discussion yesterday I thought we wanted to keep it building with 16.04 because of our large remaining userbase there. So we'd need to re-introduce that Github Action chunk and make it behave.
> 
> That should make it easier to review as well as backport.
> For stable-4.0, I'm tempted to have us offer both. That way we don't regress current build scripts but we also have meson for those who prefer it.

We shouldn't backport this to stable. This is just an invitation for breakage and the test scripts can't support both build systems and I'm very unenthusiastic about hacking them to do so as this is guaranteed to cause mayhem and breakage.

**Comment by brauner** [comment]
h

> Looks pretty good, thanks!
> 
> Can you split:
> 
> * Code refactoring (I see some defines, includes, chunks, ... moving around)
> * Code changes needed because of stricter build flags
> * Removal of autoconf/automake
> * Addition of meson
> * Porting of CI scripts

These steps can't really be switched since this will leave LXCFS in a broken state in between where it can't be compiled. We need to switch in one go. So we should switch to Meson and port the scripts in one commit. Then remove the old build system.

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+endforeach
+
+conf.set('_FILE_OFFSET_BITS', 64)
+
+libfuse = cc.find_library('fuse3', required : false)
+if libfuse.found()
+	conf.set10('HAVE_FUSE3', libfuse.found())
+	conf.set('FUSE_USE_VERSION', 30)
+else
+	libfuse = cc.find_library('fuse')
```
Comment (eli-schwartz) [inline meson.build:134]: Fuse and fuse3 have pkg-config files, fuse.pc and fuse3.pc

I'm not really sure what/how fuse works, but FUSE_USE_VERSION seems like an API level for which the ability to do

```
dependency('fuse', version: '>= 2.6')
```

May be useful?

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	conf.set10('HAVE_FUSE', libfuse.found())
```
Comment (eli-schwartz) [inline meson.build:135]: It doesn't seem this define is ever used, it is just detected by virtue of the "else" clause in variious HAVE_FUSE3 checks.

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	'src/sysfs_fuse.c',
+	'src/sysfs_fuse.h',
+	'src/utils.c',
+	'src/utils.h')
+
+liblxcfs = shared_library(
+        'lxcfs',
+        liblxcfs_sources,
+        link_args : ['-shared'],
+        dependencies : [threads,
```
Comment (eli-schwartz) [inline meson.build:199]: -shared is already handled by the shared_library() function. Given the dependencies and liblxcfs_sources are shared between two targets, you could use:

```
commondep = declare_dependency(
    sources: liblxcfs_sources,
    dependencies: [threads, libdl, libfuse])
```

and then just have both shared_library() add a dependencies: commondep

File: meson.build (hunk 1 above)
Comment (eli-schwartz) [reply meson.build:134]: Also https://mesonbuild.com/Wrap-dependency-system-manual.html

You can add a small file at subprojects/fuse3.wrap:

```
[wrap-file]
directory = libfuse-fuse-3.10.5
source_url = https://github.com/libfuse/libfuse/archive/refs/tags/fuse-3.10.5.tar.gz
source_filename = fuse-3.10.5.tar.gz
source_hash = e73f75e58da59a0e333d337c105093c496c0fd7356ef3a5a540f560697c9c4e6

[provide]
fuse3 = libfuse_dep
```

and meson will automatically download and build a copy of fuse3 if it cannot find one on the system.

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+cc = meson.get_compiler('c')
+meson_build_sh = find_program('tools/meson-build.sh')
+
+want_tests = get_option('tests')
+
+possible_cc_flags = [
+	'-fPIE',
+	'-Wvla',
+	'-std=gnu11',
```
[the remark this answers is not in the document]
Comment (brauner) [reply meson.build:54]: done

File: meson.build (hunk 1 above)
Comment (brauner) [reply meson.build:134]: Hm, the problem with this is that `dependency('fuse')` doesn't look for `fuse3` when just specifying `fuse`. No big deal. I'll work around it.

File: meson.build (hunk 2 above)
Comment (brauner) [reply meson.build:135]: I reworked this, thanks!

File: meson.build (hunk 3 above)
Comment (brauner) [reply meson.build:199]: Thanks, done!

**Pull request merged**